File: UEReflectWatch/Buildrunner.cs
using System;
using System.Diagnostics;
using System.Threading.Tasks;
using Microsoft.VisualStudio.Shell;
using Microsoft.VisualStudio.Shell.Interop;

namespace UEReflectWatch
{
    public enum BuildResult { Succeeded, Failed }

    public static class BuildRunner
    {
        public static async Task<BuildResult> RunCycleAsync(
            UnrealProject project,
            IVsOutputWindowPane outputPane,
            UEReflectWatchOptions options)
        {
            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();

            Log(outputPane, string.Empty);
            Log(outputPane, "=== UE Reflect Watch: Rebuild Cycle Started ===");
            Log(outputPane, $"Project : {project.ProjectName}");
            Log(outputPane, $"Engine  : {project.EnginePath}");
            Log(outputPane, string.Empty);

            // Step 1: Kill the editor.
            Log(outputPane, "[1/3] Closing Unreal Editor...");
            bool killed = KillEditor();
            if (killed)
            {
                Log(outputPane, $"      Editor closed. Waiting {options.KillEditorGracePeriodMs}ms...");
                await Task.Delay(options.KillEditorGracePeriodMs);
            }
            else
            {
                Log(outputPane, "      Editor was not running.");
            }

            // Step 2: Build.
            Log(outputPane, "[2/3] Building...");
            var result = await BuildAsync(project, outputPane);

            if (result == BuildResult.Failed)
            {
                Log(outputPane, string.Empty);
                Log(outputPane, "=== Build FAILED. Editor will not be relaunched. ===");
                return BuildResult.Failed;
            }

            Log(outputPane, string.Empty);
            Log(outputPane, "[3/3] Build succeeded.");

            // Step 3: Relaunch editor.
            if (options.AutoRelaunchEditor)
            {
                Log(outputPane, "      Launching Unreal Editor...");
                LaunchEditor(project);
                Log(outputPane, "      Editor launched.");
            }
            else
            {
                Log(outputPane, "      Auto-relaunch is disabled. Open the editor manually.");
            }

            Log(outputPane, string.Empty);
            Log(outputPane, "=== UE Reflect Watch: Rebuild Cycle Complete ===");

            return BuildResult.Succeeded;
        }

        private static bool KillEditor()
        {
            var processes = Process.GetProcessesByName("UnrealEditor");
            if (processes.Length == 0) return false;

            foreach (var proc in processes)
            {
                try { proc.Kill(); proc.WaitForExit(5000); }
                catch { }
                finally { proc.Dispose(); }
            }
            return true;
        }

        private static async Task<BuildResult> BuildAsync(UnrealProject project, IVsOutputWindowPane outputPane)
        {
            var args = $"\"{project.ProjectName}Editor\" Win64 Development " +
                       $"-Project=\"{project.UprojectPath}\" -WaitMutex";

            var startInfo = new ProcessStartInfo
            {
                FileName = project.BuildBatPath,
                Arguments = args,
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true
            };

            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
            Log(outputPane, $"      {project.BuildBatPath} {args}");
            Log(outputPane, string.Empty);

            using var proc = new Process { StartInfo = startInfo };

            proc.OutputDataReceived += (_, e) =>
            {
                if (e.Data is null) return;
                ThreadHelper.JoinableTaskFactory.Run(async () =>
                {
                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
                    Log(outputPane, $"      {e.Data}");
                });
            };

            proc.ErrorDataReceived += (_, e) =>
            {
                if (e.Data is null) return;
                ThreadHelper.JoinableTaskFactory.Run(async () =>
                {
                    await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
                    Log(outputPane, $"      ERR: {e.Data}");
                });
            };

            proc.Start();
            proc.BeginOutputReadLine();
            proc.BeginErrorReadLine();

            await Task.Run(() => proc.WaitForExit());

            return proc.ExitCode == 0 ? BuildResult.Succeeded : BuildResult.Failed;
        }

        private static void LaunchEditor(UnrealProject project)
        {
            try
            {
                Process.Start(new ProcessStartInfo
                {
                    FileName = project.EditorExePath,
                    Arguments = $"\"{project.UprojectPath}\"",
                    UseShellExecute = true
                });
            }
            catch (Exception ex)
            {
                Debug.WriteLine($"[UEReflectWatch] Failed to launch editor: {ex.Message}");
            }
        }

        private static void Log(IVsOutputWindowPane pane, string message)
        {
            ThreadHelper.ThrowIfNotOnUIThread();
            pane.OutputStringThreadSafe(message + Environment.NewLine);
        }
    }
}

File: RiderPlugin/src/main/kotlin/com/ueReflectWatch/runner/BuildRunner.kt
package com.ueReflectWatch.runner

import com.intellij.openapi.diagnostic.Logger
import com.ueReflectWatch.resolver.UnrealProject
import com.ueReflectWatch.settings.UEReflectWatchSettings
import java.io.File

private val LOG = Logger.getInstance("UEReflectWatch.BuildRunner")

enum class BuildResult { Succeeded, Failed }

data class BuildCycleResult(
        val result: BuildResult,
        val output: List<String>
)

object BuildRunner {

    fun runCycle(project: UnrealProject, onOutput: (String) -> Unit): BuildCycleResult {
        val settings = UEReflectWatchSettings.instance
        val output = mutableListOf<String>()

        fun log(line: String) {
            output.add(line)
            onOutput(line)
            LOG.info(line)
        }

        log("")
        log("=== UE Reflect Watch: Rebuild Cycle Started ===")
        log("Project : ${project.projectName}")
        log("Engine  : ${project.enginePath}")
        log("")

        // Step 1: Kill the editor.
        log("[1/3] Closing Unreal Editor...")
        try {
            project.adapter.killEditor()
            log("      Editor closed. Waiting ${settings.killEditorGracePeriodMs}ms...")
            Thread.sleep(settings.killEditorGracePeriodMs.toLong())
        } catch (e: Exception) {
            log("      Editor was not running or could not be closed: ${e.message}")
        }

        // Step 2: Build.
        log("[2/3] Building...")
        val buildResult = build(project, ::log)

        if (buildResult == BuildResult.Failed) {
            log("")
            log("=== Build FAILED. Editor will not be relaunched. ===")
            return BuildCycleResult(BuildResult.Failed, output)
        }

        log("")
        log("[3/3] Build succeeded.")

        // Step 3: Relaunch editor.
        if (settings.autoRelaunchEditor) {
            log("      Launching Unreal Editor...")
            try {
                project.adapter.launchEditor(project.enginePath, project.uprojectPath)
                log("      Editor launched.")
            } catch (e: Exception) {
                log("      Failed to launch editor: ${e.message}")
            }
        } else {
            log("      Auto-relaunch is disabled. Open the editor manually.")
        }

        log("")
        log("=== UE Reflect Watch: Rebuild Cycle Complete ===")

        return BuildCycleResult(BuildResult.Succeeded, output)
    }

    private fun build(
            project: UnrealProject,
            log: (String) -> Unit
    ): BuildResult {
        val args = listOf(
                project.buildScriptPath,
                "${project.projectName}Editor",
                project.buildPlatformArg,
                "Development",
                "-Project=${project.uprojectPath}",
                "-WaitMutex"
        )

        log("      ${args.joinToString(" ")}")
        log("")

        return try {
            val process = ProcessBuilder(args)
                    .redirectErrorStream(true)
                    .directory(File(project.enginePath))
                    .start()

            process.inputStream.bufferedReader().forEachLine { line ->
                log("      $line")
            }

            process.waitFor()

            if (process.exitValue() == 0) BuildResult.Succeeded else BuildResult.Failed
        } catch (e: Exception) {
            log("      Process error: ${e.message}")
            BuildResult.Failed
        }
    }
}